File: frontend/src/types/investigation.ts
/**
 * Investigations and the execution-timeline events that narrate them.
 *
 * Mirrors `backend/schemas/investigation.py` for the REST shapes and
 * `backend/api/v1/stream.py` + `docs/api-reference.md` §5 for the SSE shapes. The two
 * describe the same run from different angles and must not be conflated: `GET
 * /investigations/{id}` is a *snapshot* the client may poll, the stream is the *live
 * view*, and `docs/frontend.md` §3.2 makes the stream primary with the snapshot as the
 * fallback on reconnect exhaustion and on late attach.
 *
 * Three divergences from the prose contract are encoded here on purpose, because a
 * client that trusts the prose will render nonsense against the built backend.
 *
 * **The state vocabulary is the persisted one.** §4.2 draws a state machine over
 * `queued → planning → collecting → retrieving → analyzing → reflecting → reporting →
 * completed` plus `timed_out`. `models/enums.py::InvestigationStatus` — the column that
 * actually holds this value — is coarser: there is nowhere to store `collecting` as
 * distinct from `retrieving`, and `timed_out` is recorded as `failed` with a reason.
 * `InvestigationState` below is the stored vocabulary. A progress UI keyed to the
 * documented one would sit on "collecting" forever.
 *
 * **Several documented counters are always null.** `counts.signals_considered`,
 * `counts.evidence`, `usage.tool_calls`, and a step's `evidence_count` and `tool_calls`
 * have no column and no source. They are `null`, not `0`, and rendering them as `0`
 * reports every investigation as having considered no evidence.
 *
 * **`depth` is null on a read.** It is accepted at creation and echoed in the `202`, but
 * `investigations` has no depth column, so `GET` cannot recover it. Defaulting the read
 * to `standard` in the UI would claim a budget the run may never have been given.
 */

import type { Open, Page, TimeWindow, Usage } from '@/types/api';
import type { Platform } from '@/types/signal';

/**
 * Terminal and non-terminal states of a run — `models/enums.py::InvestigationStatus`.
 *
 * `completed_with_findings` is not a failure. It is the terminal state when the Critic
 * loop hit `MAX_CRITIC_REVISIONS` without reaching an `accept` verdict
 * (`docs/agent-system.md` §13): the report ships, with its unresolved findings surfaced
 * rather than withheld or silently presented as clean. A UI that styles it like `failed`
 * hides a usable report; one that styles it like `completed` hides the caveat.
 */
export type InvestigationState = Open<
  | 'queued'
  | 'planning'
  | 'running'
  | 'reflecting'
  | 'completed'
  | 'completed_with_findings'
  | 'failed'
  | 'cancelled'
  | 'unknown'
>;

/** States after which nothing further will happen — and after which the stream closes. */
export const TERMINAL_INVESTIGATION_STATES = [
  'completed',
  'completed_with_findings',
  'failed',
  'cancelled',
] as const satisfies readonly InvestigationState[];

/**
 * Whether a run has reached a state it will never leave.
 *
 * Used to stop polling and, more importantly, to refuse to reconnect the stream. §5 is
 * explicit that the server closes after a terminal event and that clients must not
 * reconnect — an auto-reconnect on a terminal event is a client bug that presents as a
 * server loop, and it is the single most expensive mistake available to this page.
 */
export function isTerminalState(state: InvestigationState): boolean {
  return (TERMINAL_INVESTIGATION_STATES as readonly string[]).includes(state);
}

/** Lifecycle of one agent execution inside a run (`models/orm/investigation.py`). */
export type StepState = Open<
  'pending' | 'running' | 'completed' | 'failed' | 'skipped' | 'cancelled' | 'unknown'
>;

/** The ten agents of Design Doc §9 — `models/enums.py::AgentName`. */
export type AgentName = Open<
  | 'planner'
  | 'collector'
  | 'retriever'
  | 'trend'
  | 'competitor'
  | 'forecast'
  | 'insight'
  | 'strategy'
  | 'critic'
  | 'report'
  | 'unknown'
>;

/** Budget preset chosen at creation (§4.1). Not persisted — see the module docstring. */
export type InvestigationDepth = 'quick' | 'standard' | 'deep';

/** What the investigation is allowed to look at (§4.1). */
export interface InvestigationScope {
  /** Connector slugs. **Empty means every enabled connector**, not "none". */
  platforms?: Platform[];
  /** Entity ids or names to anchor retrieval on. Empty means unanchored. */
  entities?: string[];
  /** ISO 639-1 codes. Empty means all. */
  languages?: string[];
  time_window?: TimeWindow;
}

/** Hard ceilings on spend for one run. `null` means the deployment default (§4.1). */
export interface InvestigationBudget {
  max_tokens?: number | null;
  max_tool_calls?: number | null;
}

/** The `POST /api/v1/investigations` body (§4.1). */
export interface CreateInvestigationRequest {
  /** 1–2000 characters. A whitespace-only question is an empty question. */
  query: string;
  /** What the answer will be used for. Steers the Strategy agent. */
  objective?: string | null;
  depth?: InvestigationDepth;
  scope?: InvestigationScope;
  /** Run an incremental sync before retrieval. Increases latency measurably. */
  refresh_connectors?: boolean;
  /** 1–200. Hard stop for the orchestrator. */
  max_steps?: number | null;
  budget?: InvestigationBudget;
  /** HTTPS only — the backend rejects `http://` at the edge. */
  callback_url?: string | null;
  /** ≤16 keys, values ≤256 characters. Echoed back verbatim. */
  metadata?: Record<string, string>;
}

/**
 * Where to go next. Relative paths, never absolute URLs.
 *
 * The API is reached through a proxy, a tunnel and a browser origin the backend process
 * cannot see, so it declines to guess a scheme or a hostname. That makes these safe to
 * hand to `fetch` through the same client that produced them, and unsafe to render as an
 * external link.
 */
export interface InvestigationLinks {
  self: string;
  stream: string;
  report: string | null;
}

/** The `202 Accepted` body of §4.1. Never a `201` — every creating endpoint is async. */
export interface InvestigationCreated {
  id: string;
  state: InvestigationState;
  query: string;
  depth: InvestigationDepth;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
  /**
   * Allocated eagerly so a client can subscribe before the report exists. Fetching it
   * before the run reaches reporting returns `409 report_not_ready` with a `retry_after`.
   */
  report_id: string | null;
  trace_id: string;
  links: InvestigationLinks;
}

/**
 * How far along the run is.
 *
 * `steps_total_estimate` is an estimate and not a denominator. The Critic loop can send
 * the graph round again (`docs/agent-system.md` §13), so the plan's step count is a lower
 * bound; a progress bar that divides by it will visibly run backwards mid-investigation.
 * It is `null` before the Planner has recorded a decomposition.
 */
export interface InvestigationProgress {
  steps_completed: number;
  steps_total_estimate: number | null;
}

/** Volume of what the run has produced. Nulls are honest, not zeroes — see the docstring. */
export interface InvestigationCounts {
  signals_considered: number | null;
  evidence: number | null;
  citations: number | null;
}

/**
 * Why a terminal run ended badly. Present only when `state` is `failed` or `cancelled`.
 *
 * A failed investigation is still a `200`: the request succeeded, the investigation did
 * not. Conflating the two makes a client retry an HTTP call that worked.
 *
 * `code` and `message` carry the same string today — `investigations.error` holds the
 * `code` of the raised `OmniSenseError` rather than a stack trace, precisely so that
 * handlers never pattern-match on message text.
 */
export interface InvestigationError {
  code: string;
  message: string;
  step_id: string | null;
}

/** One row of the step sub-collection (§4.2). */
export interface StepItem {
  id: string;
  /** Strictly increasing within an investigation. The stream's `seq` is a *different* counter. */
  seq: number;
  agent: AgentName;
  title: string;
  state: StepState;
  started_at: string | null;
  completed_at: string | null;
  duration_ms: number | null;
  /** Always null today: `investigation_steps` has no such column. */
  tool_calls: number | null;
  /** Always null today: there is no evidence table to count from. */
  evidence_count: number | null;
}

/** The `200` body of `GET /api/v1/investigations/{id}` (§4.2). */
export interface InvestigationDetail {
  id: string;
  state: InvestigationState;
  query: string;
  /** Null on a read. See the module docstring — do not default it to `standard`. */
  depth: InvestigationDepth | null;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
  progress: InvestigationProgress;
  counts: InvestigationCounts;
  report_id: string | null;
  trace_id: string;
  error: InvestigationError | null;
  links: InvestigationLinks;

  /**
   * The next three are **omitted**, not nulled, when `include` did not ask for them.
   * Omission is what keeps "you did not request this" distinguishable from "this is
   * empty", which matters most for `plan`: a null plan means *not planned yet*, and a
   * response that nulled an unrequested plan would throw that distinction away at the
   * last hop.
   */
  plan?: Record<string, unknown> | null;
  steps?: Page<StepItem>;
  usage?: Usage;
}

/** Sub-resources `GET /investigations/{id}` will materialise on request (§4.2). */
export type InvestigationInclude = 'plan' | 'steps' | 'evidence' | 'usage';

/** Query parameters for `GET /api/v1/investigations/{id}`. */
export interface InvestigationQuery {
  include?: InvestigationInclude[];
  /** Max 200. Values above the max are rejected with 422, never clamped. */
  steps_limit?: number;
  steps_cursor?: string;
}

// --------------------------------------------------------------------------- //
// The execution timeline (SSE, §5)
// --------------------------------------------------------------------------- //

/**
 * Event names carried by `GET /investigations/{id}/stream`.
 *
 * `stream.gap` is an OmniSense extension emitted by `backend/api/v1/stream.py` when a
 * range of events will never arrive — the resume point fell out of the replay buffer, or
 * backpressure dropped frames. It is deliberately sent **without an `id:` line** so that
 * a client's `Last-Event-ID` does not advance past events it never received; advancing it
 * would make the gap permanent across every future reconnect.
 *
 * Open, because §1 makes adding an event type additive and requires clients to ignore
 * names they do not know.
 */
export type TimelineEventType = Open<
  | 'step.started'
  | 'tool.called'
  | 'evidence.found'
  | 'step.completed'
  | 'error'
  | 'done'
  | 'stream.gap'
>;

/**
 * The envelope every `data` payload shares.
 *
 * `seq` is assigned by the orchestrator — the only party that sees every event of a run
 * in order — and is strictly increasing per investigation. A gap in `seq` means a
 * *dropped* event, never a reordered one, so the correct response is to surface it and
 * fall back to the snapshot, not to interpolate the missing rows.
 */
export interface TimelineEnvelope {
  investigation_id: string;
  seq: number;
  ts: string;
  /** Correlation id, present on every frame so an operator can grep one connection. */
  request_id?: string | null;
}

export interface StepStartedEvent extends TimelineEnvelope {
  type: 'step.started';
  step_id: string;
  agent: AgentName;
  title: string;
  parent_step_id?: string | null;
  plan_index?: number | null;
}

/**
 * An agent invoking a tool. Emitted twice for one call: once `started`, once terminal.
 *
 * `arguments` is redacted by the server for any tool holding credentials, so it may be
 * absent or partial and must never be assumed complete. A `failed` tool call does not
 * imply a failed step — the router can route around one.
 */
export interface ToolCalledEvent extends TimelineEnvelope {
  type: 'tool.called';
  step_id: string;
  tool: string;
  arguments?: Record<string, unknown> | null;
  duration_ms: number | null;
  status: Open<'started' | 'completed' | 'failed'>;
  error?: string | null;
}

/**
 * Evidence admitted by the Retriever or Collector.
 *
 * Payloads are deliberately small — ids and a snippet. Full evidence is hydrated on
 * demand through `GET /signals`, which is why `signal_id` is here and the signal body is
 * not. `signal_id` is `sig_`-prefixed and opaque; validating it as a UUID rejects every
 * real one.
 */
export interface EvidenceFoundEvent extends TimelineEnvelope {
  type: 'evidence.found';
  step_id: string;
  evidence_id: string;
  signal_id: string;
  platform: Platform;
  url: string | null;
  snippet: string;
  /** A retrieval score. Never rendered as confidence. */
  retrieval_score: number;
  retriever: Open<'vector' | 'keyword' | 'graph' | 'hybrid'>;
}

export interface StepCompletedEvent extends TimelineEnvelope {
  type: 'step.completed';
  step_id: string;
  agent: AgentName;
  state: Open<'completed' | 'skipped' | 'failed'>;
  duration_ms: number | null;
  evidence_count: number | null;
  tool_calls: number | null;
  usage?: { input_tokens: number; output_tokens: number } | null;
  summary?: string | null;
}

/** Terminal. The investigation itself failed; `code` is the §3.3 vocabulary. */
export interface TimelineErrorEvent extends TimelineEnvelope {
  type: 'error';
  code: string;
  message: string;
  step_id: string | null;
  retryable: boolean;
}

/** Terminal. Always the last event of a run that reached a terminal state. */
export interface DoneEvent extends TimelineEnvelope {
  type: 'done';
  state: InvestigationState;
  report_id: string | null;
  duration_ms: number | null;
  counts?: { steps: number; evidence: number; citations: number } | null;
  usage?: Usage | null;
}

/** OmniSense extension: events in `[from_seq, to_seq]` will never be delivered here. */
export interface StreamGapEvent extends TimelineEnvelope {
  type: 'stream.gap';
  from_seq: number;
  to_seq: number;
  reason: string;
}

/**
 * An event whose name this bundle does not recognise.
 *
 * Required by §1, not defensive programming: a backend deployed ahead of this bundle may
 * publish a name added after it was built, and the contract says to ignore it. Typing it
 * keeps "ignore" a compile-time-checked branch instead of a silent `default:`.
 */
export interface UnknownTimelineEvent extends TimelineEnvelope {
  type: Open<string>;
  [key: string]: unknown;
}

export type TimelineEvent =
  | StepStartedEvent
  | ToolCalledEvent
  | EvidenceFoundEvent
  | StepCompletedEvent
  | TimelineErrorEvent
  | DoneEvent
  | StreamGapEvent
  | UnknownTimelineEvent;

/** Event names after which the server closes and the client must not reconnect (§5). */
export const TERMINAL_EVENT_TYPES = ['error', 'done'] as const;

export function isTerminalEvent(event: TimelineEvent): boolean {
  return event.type === 'error' || event.type === 'done';
}

/**
 * The timeline as the UI holds it: steps in `seq` order, each with its tool calls and
 * evidence nested underneath.
 *
 * Derived from the event log rather than fetched, and stored in the TanStack Query cache
 * beside the investigation itself. It is *not* mirrored into Zustand: a second copy of
 * the timeline is the defect `docs/frontend.md` §3.2 names explicitly, because the two
 * copies diverge the first time a reconnect replays an event the store already applied.
 */
export interface TimelineStep {
  step_id: string;
  seq: number;
  agent: AgentName;
  title: string;
  state: StepState;
  started_at: string | null;
  completed_at: string | null;
  duration_ms: number | null;
  summary: string | null;
  tool_calls: ToolCallRecord[];
  evidence: EvidenceFoundEvent[];
}

/** One tool invocation, folded from its `started` and terminal `tool.called` events. */
export interface ToolCallRecord {
  tool: string;
  status: Open<'started' | 'completed' | 'failed'>;
  arguments: Record<string, unknown> | null;
  duration_ms: number | null;
  error: string | null;
  /** `seq` of the frame that opened the call; the sort key that keeps calls in order. */
  seq: number;
}

File: backend/schemas/investigation.py
"""Wire DTOs for `POST /investigations` and `GET /investigations/{id}`.

`docs/api-reference.md` §4.1 and §4.2 are the contract. Two places where this
module deliberately does **not** match the document are called out here rather
than buried, because both are real divergences a client will notice and neither
is fixable at the schema layer.

**1. The state vocabulary is the persisted one, not the document's.**
§4.2 draws a state machine over `queued -> planning -> collecting -> retrieving ->
analyzing -> reflecting -> reporting -> completed`, plus `timed_out`. The column
that holds this value is `investigations.status`, typed
`TolerantEnumType(InvestigationStatus)` (`models/orm/investigation.py`), and
`models/enums.py::InvestigationStatus` has a coarser vocabulary: `queued`,
`planning`, `running`, `reflecting`, `completed`, `completed_with_findings`,
`failed`, `cancelled`. There is nowhere to store `collecting` as distinct from
`retrieving`, and `timed_out` is recorded as `failed` with a reason.

Inventing a mapping here would be worse than the divergence. Reporting
`retrieving` for a row that says `running` would make the API assert something
the database does not know, and `GET` would contradict the SSE stream, the
orchestrator's own checkpoint and the reaper. So the wire carries the stored
value. §1's versioning policy explicitly permits this direction -- *"Adding a new
enum member to a response field: yes, clients must tolerate unknown members"* --
and `completed_with_findings` is a state clients genuinely need
(`docs/agent-system.md` §13: the report ships with unresolved Critic findings
surfaced rather than withheld). Closing the gap properly means finer statuses in
`models/enums.py` and a migration, not a translation table in a DTO.

**2. Several documented counters have no column and are reported as `null`.**
`counts.signals_considered`, `counts.evidence`, `usage.tool_calls`, and a step's
`evidence_count` and `tool_calls` are all in §4.2's example. None of them exists:
there is no `evidence` table in `models/orm/`, `investigation_steps` has no
tool-call column, and nothing records how many Signals a retrieval pass looked at.

`null` and not `0`. This repository already takes that position where it matters
-- `Engagement.compute_score` returns `None` rather than `0.0` because "nobody
engaged" and "we did not measure" are different claims, and a dashboard cannot
tell them apart once the second has been rendered as the first. A `0` here would
show every investigation as having considered no evidence.
"""

from __future__ import annotations

import enum
from datetime import datetime
from typing import Any

from pydantic import Field, field_validator, model_validator

from backend.schemas.common import Page, RequestModel, ResponseModel
from models.enums import AgentName, InvestigationStatus
from models.orm.investigation import StepStatus

__all__ = [
    "MAX_METADATA_KEYS",
    "MAX_METADATA_VALUE_CHARS",
    "MAX_QUERY_CHARS",
    "CreateInvestigationRequest",
    "InvestigationBudget",
    "InvestigationCounts",
    "InvestigationCreated",
    "InvestigationDepth",
    "InvestigationDetail",
    "InvestigationError",
    "InvestigationLinks",
    "InvestigationProgress",
    "InvestigationScope",
    "InvestigationUsage",
    "StepItem",
    "StepsPage",
    "TimeWindow",
]

MAX_QUERY_CHARS = 2000
"""§4.1: "1-2000 chars".

Restated from `services/investigation_service.py`, which enforces the same bound
at the service boundary. Both are needed: this one gives the caller a `422`
naming the field, and the service's protects the `Text` column from a caller that
never came through HTTP.
"""

MAX_METADATA_KEYS = 16
MAX_METADATA_VALUE_CHARS = 256
"""§4.1: "<=16 keys, string values <=256 chars; echoed back verbatim".

Bounded because it *is* echoed back verbatim, which makes it the one field a
caller controls end to end. Unbounded, it is free storage attached to every
investigation and a reflection surface in every response built from one.
"""


class InvestigationDepth(enum.StrEnum):
    """Budget preset from §4.1. Not persisted -- see `CreateInvestigationRequest`."""

    QUICK = "quick"
    STANDARD = "standard"
    DEEP = "deep"


class TimeWindow(RequestModel):
    """Half-open bound on signal timestamps: `[from, to)`.

    `from` is a Python keyword, so the field is `from_` with an alias. The
    half-open convention is inherited from `services/signal_service.py`, where it
    is load-bearing: consecutive daily windows tile without overlap, so a Signal
    sitting exactly on midnight is neither double-counted nor missed, and trend
    volume is computed from those counts.
    """

    from_: datetime | None = Field(default=None, alias="from")
    to: datetime | None = None

    @field_validator("from_", "to")
    @classmethod
    def _must_be_aware(cls, value: datetime | None) -> datetime | None:
        """§3.2: "Naive datetimes are rejected with 422".

        Enforced rather than defaulted because the column being compared against
        is `TIMESTAMP WITH TIME ZONE`: a naive bound is compared against whatever
        timezone the driver assumed, which silently shifts the window by hours and
        produces a plausible, wrong answer.
        """
        if value is not None and value.tzinfo is None:
            raise ValueError("must be timezone-aware, e.g. 2026-05-01T00:00:00Z")
        return value

    @model_validator(mode="after")
    def _ordered(self) -> TimeWindow:
        if self.from_ is not None and self.to is not None and self.from_ > self.to:
            raise ValueError("`from` is later than `to`; the window is empty")
        return self


class InvestigationScope(RequestModel):
    """What the investigation is allowed to look at (§4.1).

    Every list defaults to empty, and empty means *unrestricted* rather than
    *nothing*: "empty means every enabled connector", "empty means all". That
    reading is stated in §4.1 for `platforms` and `languages`, and it is the only
    coherent one -- a default that matched nothing would make the simplest
    possible request return no evidence.
    """

    platforms: list[str] = Field(default_factory=list, max_length=64)
    entities: list[str] = Field(default_factory=list, max_length=64)
    languages: list[str] = Field(default_factory=list, max_length=32)
    time_window: TimeWindow = Field(default_factory=TimeWindow)


class InvestigationBudget(RequestModel):
    """Hard ceilings on spend for one run (§4.1). `None` means the deployment default."""

    max_tokens: int | None = Field(default=None, ge=1000)
    max_tool_calls: int | None = Field(default=None, ge=1)


class CreateInvestigationRequest(RequestModel):
    """The `POST /investigations` body.

    Most of this is accepted and validated but not yet *persisted*, and the
    reason is structural rather than an omission: `investigations` has columns for
    the question, the plan, the status and the accounting, and none for scope,
    depth, budget, callback or metadata (`models/orm/investigation.py`). Those
    belong to the orchestrator's run configuration, and the orchestrator
    (`agents/graph.py` and the worker that drives it) is what will consume them.

    Validating them here anyway is the point of the endpoint being a contract:
    a caller that sends `from` after `to`, a naive datetime, seventeen metadata
    keys or `max_steps=0` gets the documented `422` today, and the day the
    orchestrator reads these fields nothing about the request changes.
    """

    query: str = Field(min_length=1, max_length=MAX_QUERY_CHARS)
    objective: str | None = Field(default=None, max_length=2000)
    depth: InvestigationDepth = InvestigationDepth.STANDARD
    scope: InvestigationScope = Field(default_factory=InvestigationScope)
    refresh_connectors: bool = False
    max_steps: int | None = Field(default=None, ge=1, le=200)
    budget: InvestigationBudget = Field(default_factory=InvestigationBudget)
    callback_url: str | None = Field(default=None, max_length=2048)
    metadata: dict[str, str] = Field(default_factory=dict)

    @field_validator("query")
    @classmethod
    def _query_is_not_blank(cls, value: str) -> str:
        """A whitespace-only question is an empty question.

        `RequestModel` strips strings, but relying on the strip alone couples this
        rule to Pydantic's field-processing order. Checking the stripped value
        explicitly makes the outcome independent of it.
        """
        if not value.strip():
            raise ValueError("an investigation needs a question")
        return value

    @field_validator("callback_url")
    @classmethod
    def _https_only(cls, value: str | None) -> str | None:
        """§4.1: "HTTPS only".

        A webhook carries the investigation's outcome to a URL the caller chose,
        so plain HTTP would put that outcome on the wire in clear text to a host
        this system does not authenticate. Rejected at the edge because the field
        is stored and used later, by a process with no caller left to ask.
        """
        if value is None:
            return None
        if not value.startswith("https://"):
            raise ValueError("callback_url must be an https:// URL")
        return value

    @field_validator("metadata")
    @classmethod
    def _bounded_metadata(cls, value: dict[str, str]) -> dict[str, str]:
        if len(value) > MAX_METADATA_KEYS:
            raise ValueError(f"metadata accepts at most {MAX_METADATA_KEYS} keys")
        for key, item in value.items():
            if len(item) > MAX_METADATA_VALUE_CHARS:
                raise ValueError(
                    f"metadata[{key!r}] is longer than {MAX_METADATA_VALUE_CHARS} characters"
                )
        return value


class InvestigationLinks(ResponseModel):
    """Where to go next. §4.1 returns these alongside the id.

    Relative paths, not absolute URLs. The API is reached through a proxy, a
    tunnel and a browser origin that this process cannot see -- `Host` is
    forgeable and `X-Forwarded-*` is only as trustworthy as the last hop -- so
    building an absolute URL here would eventually hand a client a link to the
    wrong scheme or to an internal hostname.
    """

    self: str
    stream: str
    report: str | None = None


class InvestigationCreated(ResponseModel):
    """The `202` body of §4.1."""

    id: str
    state: InvestigationStatus
    query: str
    depth: InvestigationDepth
    created_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None
    report_id: str | None = Field(
        default=None,
        description="Allocated eagerly so a client can subscribe before the report "
        "exists. Fetching it before the investigation reaches reporting returns "
        "409 report_not_ready.",
    )
    trace_id: str
    links: InvestigationLinks


class StepItem(ResponseModel):
    """One row of the step sub-collection (§4.2).

    `tool_calls` and `evidence_count` are `null` for every step, always:
    `investigation_steps` has neither column, and there is no evidence table to
    count from. See this module's docstring for why they are `null` and not `0`.
    """

    id: str
    seq: int
    agent: AgentName
    title: str
    state: StepStatus
    started_at: datetime | None = None
    completed_at: datetime | None = None
    duration_ms: int | None = None
    tool_calls: int | None = None
    evidence_count: int | None = None


type StepsPage = Page[StepItem]
"""The paginated step sub-collection. Cursor-based like every other collection."""


class InvestigationProgress(ResponseModel):
    """How far along the run is (§4.2).

    `steps_total_estimate` comes from the Planner's decomposition when one has
    been recorded, and is `null` before that. It is named an *estimate* in the
    contract for a good reason: the Critic loop can send the graph round again
    (`docs/agent-system.md` §13), so the plan's step count is a lower bound, not a
    denominator. A UI that renders it as one will show progress going backwards.
    """

    steps_completed: int
    steps_total_estimate: int | None = None


class InvestigationCounts(ResponseModel):
    """Volume of what the run has produced (§4.2). See the module docstring."""

    signals_considered: int | None = None
    evidence: int | None = None
    citations: int | None = None


class InvestigationUsage(ResponseModel):
    """Token and tool accounting (§4.2).

    Token counts are real: they are rolled up into `investigations.token_input`
    and `token_output` inside the same transaction that completes each step
    (`services/investigation_service.py`), precisely so the budget check on the
    next step reads a total that is not lagging a sweep interval behind.
    """

    input_tokens: int
    output_tokens: int
    tool_calls: int | None = None


class InvestigationError(ResponseModel):
    """Why a terminal run ended badly (§4.2).

    Present only when the run failed or was cancelled. §4.2 is explicit that this
    is still a `200`: the request succeeded, the investigation did not, and
    conflating the two would make a client retry an HTTP call that worked.

    `code` and `message` carry the same string today. `investigations.error` is
    documented to hold the `code` of the raised `OmniSenseError` rather than a
    stack trace, so that handlers never pattern-match on message text
    (`docs/coding-standards.md` §2.7) -- there is one column and it holds the
    code. Splitting them needs a second column, not a guess at where the code
    ends.
    """

    code: str
    message: str
    step_id: str | None = None


class InvestigationDetail(ResponseModel):
    """The `200` body of §4.2.

    `plan`, `steps` and `usage` are omitted entirely rather than nulled when the
    caller did not ask for them via `include`. Omission is what makes "you did not
    request this" distinguishable from "this is empty", which matters most for
    `plan`: `investigations.plan` is nullable precisely so that *not planned yet*
    stays distinguishable from *planned to do nothing*
    (`models/orm/investigation.py`), and a response that nulled an unrequested
    plan would throw that distinction away at the last hop.
    """

    id: str
    state: InvestigationStatus
    query: str
    depth: InvestigationDepth | None = Field(
        default=None,
        description="Null on a read: `investigations` has no depth column, so the "
        "preset chosen at creation is not recoverable. Echoing `standard` would "
        "report a budget the run may never have been given.",
    )
    created_at: datetime
    started_at: datetime | None = None
    completed_at: datetime | None = None
    progress: InvestigationProgress
    counts: InvestigationCounts
    report_id: str | None = None
    trace_id: str
    error: InvestigationError | None = None
    links: InvestigationLinks

    plan: dict[str, Any] | None = None
    steps: StepsPage | None = None
    usage: InvestigationUsage | None = None
